docs/key-issues/gpus.md

```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+   libraries compose the bare minimum required to compile CUDA code.
+2. The user-mode driver: This is the libcuda.so library. This library is
+   required to actually run CUDA code.
+3. The kernel-mode driver: The nvidia.ko file. This constitutes what is
+   typically considered a "driver" in common parlance when referring to other
+   peripherals connected to a computer.
+
+CUDA developers and users generally do not need to be aware of the kernel-mode driver.
+For the rest of this section, when referring to the "driver" we will always be referring to the user-mode driver libcuda.so.
+
+The CUDA runtime library makes no forward or backwards compatibility
+guarantees, meaning that libraries that dynamically link to the CUDA runtime
+may not work correctly if they are run on a system with a different CUDA
+runtime shared library than the one they were compiled against. In this scenario,
+users are responsible for having the right CUDA runtime library installed.  As
+a result, the official CUDA recommendation is to statically link the CUDA
+runtime (see
+[here](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#recommendations-for-building-a-minor-version-compatible-library)
+and
+[here](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#distributing-the-cuda-runtime-and-libraries)
+for more information).
+
+CUDA drivers have always been backwards compatible. Any code that runs when
+some driver version X is installed will always work correctly when run with
+some newer driver version Y>X. As briefly discussed above, as of CUDA 11.0
+CUDA also promises [minor version
+compatibility](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#minor-version-compatibility)
+(MVC). This compatibility guarantees that CUDA code compiled using a certain
+version of the CUDA toolkit will be compatible with any driver version within
+the same major release. This behavior is useful because it is often easier for
+users to upgrade their CUDA runtime than it is to upgrade the driver,
+especially on shared machines. For instance, the CUDA toolkit may be installed
+using conda, while the driver library cannot be. An example of leveraging MVC
+would be compiling code against the CUDA 11.5 runtime library and then running
+it on a system with a CUDA 11.2 driver installed. However there are some
+caveats with MVC:
+
+- If the code uses any features that were introduced in a later driver version
+  than the installed version, it will still fail to run. However, it will be a
+  runtime failure in the form of a `cudaErrorCallRequiresNewerDriver` CUDA
+  error, rather than a linker error or some similarly opaque issue. One
+  solution to this problem is for libraries to use runtime checks of the CUDA
+  version (using e.g. `cudaDriverGetVersion`) to only use supported features on
+  the installed driver.
+- NVRTC did not start supporting MVC until CUDA 11.3. Therefore, code that uses
```

Comment:
Additionally, I believe that you cannot use NVRTC to generate code that needs linking by the driver under MVC (e.g. when calling device functions from another translation unit) - it will produce a cubin, but the driver will refuse to add it to a link (with a cryptic error code, IIRC).

Reply:
Are there specific version constraints here? Is this true for any 11.x NVRTC?

Reply:
I think it's true for any NVRTC that supports MVC, including in CUDA 12 - you would still need nvJitlink for MVC with linking.

Reply:
This is my understanding:
- NVRTC -> compiler (CUDA C++ -> PTX or SASS)
- nvJitLink -> linker

and to support MVC we need both pieces, unless linking is not needed (ex: compiling a single TU) as Graham noted. 

That nvJitLink was missing in CUDA 11 is one reason that MVC could not cover a lot of PTX use cases. Another reason is that the UMD driver is responsible for compiling (JIT'ing) PTX to SASS, but if the PTX is generated with NVRTC newer than UMD (say, NVRTC from 11.6, UMD from 11.2), JIT'ing will fail, too. (Reason: Each CTK minor release bumps the PTX minor version -- yes, it has its own version -- so an old driver would simply not recognize the new code.) This was why CuPy had to compile down to SASS with NVRTC.

Reply:
Hmm @leofang [this diagram](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#deployment-considerations) indicates 11.3, not 11.2. Is it wrong?

Aside from that this discussion makes sense though, will update.

Reply:
The single TU + no linking restriction is now mentioned below.